plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpToolsRegisterOnNodesAction.java

```diff
@@ -0,0 +1,150 @@
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.toolFactoryWrapper = toolFactoryWrapper;
+    }
+
+    @Override
+    protected MLMcpRegisterNodesResponse newResponse(
+        MLMcpToolsRegisterNodesRequest nodesRequest,
+        List<MLMcpRegisterNodeResponse> responses,
+        List<FailedNodeException> failures
+    ) {
+        return new MLMcpRegisterNodesResponse(clusterService.getClusterName(), responses, failures);
+    }
+
+    @Override
+    protected MLMcpToolsRegisterNodeRequest newNodeRequest(MLMcpToolsRegisterNodesRequest request) {
+        return new MLMcpToolsRegisterNodeRequest(request.getMcpTools());
+    }
+
+    @Override
+    protected MLMcpRegisterNodeResponse newNodeResponse(StreamInput in) throws IOException {
+        return new MLMcpRegisterNodeResponse(in);
+    }
+
+    @Override
+    protected MLMcpRegisterNodeResponse nodeOperation(MLMcpToolsRegisterNodeRequest request) {
+        return registerToolsOnNode(request.getMcpTools());
+    }
+
+    private MLMcpRegisterNodeResponse registerToolsOnNode(McpTools mcpTools) {
+        Flux.fromStream(mcpTools.getTools().stream()).flatMap(tool -> {
+            // check if user request contains tools that not in our system.
+            String toolName = tool.getName();
+            Tool.Factory factory = toolFactoryWrapper.getToolsFactories().get(toolName);
+            Tool actualTool = factory.create(tool.getParams());
+            Map<String, Object> mSchema = Optional
+                .ofNullable(tool.getSchema())
+                .orElse(
+                    Optional
+                        .ofNullable(actualTool.getAttributes())
+                        .map(x -> StringUtils.fromJson(((String) x.get(INPUT_SCHEMA)), INPUT_SCHEMA))
+                        .orElse(ImmutableMap.of())
+                );
+            String schema = StringUtils.gson.toJson(mSchema);
+            String description = Optional.ofNullable(tool.getDescription()).orElse(factory.getDefaultDescription());
+            McpServerFeatures.AsyncToolSpecification toolSpecification = new McpServerFeatures.AsyncToolSpecification(
+                new McpSchema.Tool(toolName, String.valueOf(description), schema),
+                (exchange, arguments) -> Mono.create(sink -> {
+                    ActionListener<String> actionListener = ActionListener
+                        .wrap(r -> sink.success(new McpSchema.CallToolResult(List.of(new McpSchema.TextContent(r)), false)), e -> {
+                            log.error("Failed to execute tool, tool name: {}", toolName, e);
+                            sink.error(e);
+                        });
+                    actualTool.run(StringUtils.getParameterMap(arguments), actionListener);
+                })
+            );
+            return McpAsyncServerHolder.asyncServer
+                .removeTool(toolName)
```

Comment:
Re-register a tool with same name will remove the existing registered tool ?

Reply:
Yes, it's an override operation.